File: src/main/java/com/github/eendroroy/sdk/openweather/ClientFactory.java
package com.github.eendroroy.sdk.openweather;

import com.github.eendroroy.sdk.openweather.client.WeatherClient;
import com.github.eendroroy.sdk.openweather.config.DefaultWeatherConfiguration;
import com.github.eendroroy.sdk.openweather.config.WeatherConfiguration;
import com.github.eendroroy.sdk.openweather.interceptor.DefaultOWInterceptor;
import com.github.eendroroy.sdk.openweather.interceptor.OWInterceptor;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.jackson.JacksonConverterFactory;

/**
 * @author indrajit
 */

public class ClientFactory {
    private final OWInterceptor interceptor;
    private final WeatherConfiguration weatherConfiguration;

    public ClientFactory() {
        this(new DefaultWeatherConfiguration(), new DefaultOWInterceptor());
    }

    public ClientFactory(WeatherConfiguration weatherConfiguration) {
        this(weatherConfiguration, new DefaultOWInterceptor());
    }

    public ClientFactory(WeatherConfiguration weatherConfiguration, OWInterceptor interceptor) {
        this.weatherConfiguration = weatherConfiguration;
        this.interceptor = interceptor;
    }

    public WeatherClient weatherClient() {
        OkHttpClient okHttpClient = new OkHttpClient.Builder().addInterceptor(interceptor).build();

        Retrofit retrofit = new Retrofit.Builder()
                .client(okHttpClient)
                .baseUrl(weatherConfiguration.getBaseUrl())
                .addConverterFactory(JacksonConverterFactory.create())
                .build();
        return new WeatherClient(retrofit, weatherConfiguration);
    }
}

File: src/main/kotlin/com/github/eendroroy/sdk/openweathermap/ClientFactory.kt
package com.github.eendroroy.sdk.openweathermap

import com.github.eendroroy.sdk.openweathermap.client.*
import com.github.eendroroy.sdk.openweathermap.config.DefaultOWConfiguration
import com.github.eendroroy.sdk.openweathermap.config.OWConfiguration
import com.github.eendroroy.sdk.openweathermap.interceptor.DefaultOWInterceptor
import com.github.eendroroy.sdk.openweathermap.interceptor.OWInterceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.jackson.JacksonConverterFactory

/**
 * @author indrajit
 */
class ClientFactory {
    val weatherClient: WeatherClient
        get() {
            return WeatherClient(retrofit, configuration)
        }

    val hourlyForecastClient: HourlyForecastClient
        get() {
            return HourlyForecastClient(retrofit, configuration)
        }

    val dailyForecastClient: DailyForecastClient
        get() {
            return DailyForecastClient(retrofit, configuration)
        }

    val climateClient: ClimateClient
        get() {
            return ClimateClient(retrofit, configuration)
        }

    private var interceptor: OWInterceptor = DefaultOWInterceptor()
    private var configuration: OWConfiguration = DefaultOWConfiguration()

    fun with(OWConfiguration: OWConfiguration): ClientFactory {
        this.configuration = OWConfiguration
        return this
    }

    fun with(interceptor: OWInterceptor): ClientFactory {
        this.interceptor = interceptor
        return this
    }

    private val retrofit: Retrofit
        get() {
            val okHttpClientBuilder: OkHttpClient.Builder = when {
                configuration.acceptUnsafeSSL() -> {
                    UnsafeOkHttpClientBuilder.unsafeOkHttpClientBuilder
                }
                else -> {
                    OkHttpClient.Builder()
                }
            }

            val okHttpClient = okHttpClientBuilder.addInterceptor(interceptor).build()

            return Retrofit.Builder()
                    .client(okHttpClient)
                    .baseUrl(configuration.baseUrl())
                    .addConverterFactory(JacksonConverterFactory.create())
                    .build()
        }
}